### PR TITLE
Backlink queries

### PR DESCRIPTION
**Pull request by kneth**
@bmeike did all the hard work. I am merely wrapping up. This PR is still WIP!

* [x] Fix unit tests for backlink queries.
* [x] Explain what unit tests are testing.
* [x] Introduce a native implementation of isNotEmpty() - see also #2025.
* [x] ~Remove~ rewrite`getTableByArray()`.
* [x] Add support for queries when last element is a backlink.
* [x] Update documentation about link (add backlink) at web site - #4320. See also https://github.com/realm/realm.io/pull/3294

See also #4620.

### CODE REVIEW

**Comment by bmeike** [comment]
Yeah, that's like saying that thousands of years of Mathematics did all the hard work and that Laplace just kinda finished it up.  Reading now.

File: realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp [hunk 1]
```diff
@@ -1490,21 +1495,22 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeIsNull(JNIEnv* en
...
-        TableRef table_ref = getTableByArray(nativeQueryPtr, index_arr);
 
+        TableRef table_ref = getTableByArray(nativeQueryPtr, table_arr, index_arr);
         if (!isNullable(env, TBL(table_arr[arr_len - 1]), table_ref, column_idx)) {
             return;
         }
 
-        // FIXME!!!  Support a backlink as the last column in a field descriptor
-        int col_type = table_ref->get_column_type(S(column_idx));
+        // FIXME: Support a backlink as the last column in a field descriptor
```
Comment (cmelchior) [inline realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp]: I assume the behaviour would either to throw an error or just assume this will never match.

File: realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp [hunk 2]
```diff
@@ -1760,6 +1772,77 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeIsEmpty(JNIEnv* e
...
+                    pQuery->not_equal(S(column_idx), "");
+                    break;
+                case type_Link:
+                case type_Bool:
+                case type_Int:
+                case type_Float:
+                case type_Double:
+                case type_Timestamp:
+                default:
+                    ThrowException(env, IllegalArgument, "isNotEmpty() only works on String, byte[] and RealmList.");
```
Comment (cmelchior) [inline realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp]: and RealmResults ?

File: realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp [hunk 3]
```diff
@@ -1760,6 +1772,77 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeIsEmpty(JNIEnv* e
...
+                    return;
+            }
+        }
+        else {
+            // Linked queries
+            // FIXME!!!  Support a backlink as an internal column in a field descriptor
```
Comment (cmelchior) [inline realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp]: It is supported for isEmpty()` so adding support here would be trivial?

Comment (cmelchior) [inline realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp]: Also, we do have a unit test capturing this?

File: realm/realm-library/src/main/java/io/realm/internal/TableQuery.java [hunk 4]
```diff
@@ -737,6 +739,9 @@ private void throwImmutable() {
 
     private native void nativeIsEmpty(long nativePtr, long[] columnIndices, long[] tablePtrs);
 
+    private native void nativeIsNotEmpty(long nativePtr, long[] columnIndices, long[] tablePtrs);
+
+
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/internal/TableQuery.java]: Whitespace

File: realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsQueryTests.java [hunk 5]
```diff
@@ -30,7 +29,7 @@
 import static org.junit.Assert.assertTrue;
 import static org.junit.Assert.fail;
 
-@Ignore
+//@Ignore
```
Comment (bmeike) [inline realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsQueryTests.java]: I think you probably just want to delete this line.

File: realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsQueryTests.java [hunk 6]
```diff
@@ -393,13 +392,14 @@ public void isEmpty_acrossLinkingObjectObjectLink() {
     @Test
     public void isEmpty_acrossLinkingObjectListLink() {
         createIsEmptyDataSet(realm);
+        assertEquals(2, realm.where(AllJavaTypes.class).findAll().size());
         for (RealmFieldType type : SUPPORTED_IS_EMPTY_TYPES) {
             switch (type) {
                 case STRING:
-                    assertEquals(1, realm.where(AllJavaTypes.class).isEmpty(AllJavaTypes.FIELD_LO_LIST + "." + AllJavaTypes.FIELD_STRING).count());
+                    assertEquals(0, realm.where(AllJavaTypes.class).isEmpty(AllJavaTypes.FIELD_LO_LIST + "." + AllJavaTypes.FIELD_STRING).count());
```
Comment (bmeike) [inline realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsQueryTests.java]: I suspect that this test is not very robust.  There are so many ways that the query could fail to return any entries.  Is there any way to fiddle the data so that there is one thing there?

File: realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsQueryTests.java [hunk 7]
```diff
@@ -393,13 +392,14 @@ public void isEmpty_acrossLinkingObjectObjectLink() {
...
                     break;
                 case BINARY:
-                    assertEquals(1, realm.where(AllJavaTypes.class).isEmpty(AllJavaTypes.FIELD_LO_LIST + "." + AllJavaTypes.FIELD_BINARY).count());
+                    assertEquals(0, realm.where(AllJavaTypes.class).isEmpty(AllJavaTypes.FIELD_LO_LIST + "." + AllJavaTypes.FIELD_BINARY).count());
```
Comment (bmeike) [inline realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsQueryTests.java]: ... as above...

File: realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsQueryTests.java [hunk 8]
```diff
@@ -472,6 +472,7 @@ public void isNotEmpty_acrossLinkingObjectObjectLink() {
     @Test
     public void isNotEmpty_acrossLinkingObjectListLink() {
         createIsEmptyDataSet(realm);
+        assertEquals(2, realm.where(AllJavaTypes.class).findAll().size());
```
Comment (bmeike) [inline realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsQueryTests.java]: Nice.  Love it.

File: realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsQueryTests.java [hunk 9]
```diff
@@ -497,13 +498,13 @@ public void isNotEmpty_acrossLinkingObjectListLink() {
...
-    // |2| Horse  |    1 |       3 |  {1,2} |                  3 |
-    // +-+--------+------+---------+--------+--------------------+
+    // +-+--------+------+---------+--------+--------------------+----------+
+    // | | string | link | numeric | binary | numeric (not null) | linklist |
+    // +-+--------+------+---------+--------+--------------------+----------+
+    // |0| Fish   |    0 |       1 |    {0} |                  1 |      [0] |
+    // |1| null   |    2 |    null |   null |                  0 |      [2] |
+    // |2| Horse  | null |       3 |  {1,2} |                  3 |     null |
+    // +-+--------+------+---------+--------+--------------------+----------+
     private void populateTestRealmForNullTests(Realm testRealm) {
```
Comment (bmeike) [inline realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsQueryTests.java]: brilliant

File: realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp [hunk 10]
```diff
@@ -61,6 +62,31 @@ JNIEXPORT jstring JNICALL Java_io_realm_internal_TableQuery_nativeValidateQuery(
 
 // helper functions
 
+/*static TableRef resolve_link(jlong nativeQueryPtr, JniLongArray& fields) {
```
Comment (bmeike) [inline realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp]: ewwwwww!  If we don't need this, let's delete it.  Remove the evidence of my unclear thinking, please.  No commented out code.

File: realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp [hunk 11]
```diff
@@ -1504,7 +1530,7 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeIsNull(JNIEnv* en
         }
 
         // FIXME!!!  Support a backlink as the last column in a field descriptor
```
Comment (bmeike) [inline realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp]: I think this is fixed, so can we pull the "fixme"?

File: realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp [hunk 12]
```diff
@@ -1704,18 +1730,18 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeIsEmpty(JNIEnv* e
     try {
         TableRef src_table_ref = getTableForLinkQuery(nativeQueryPtr, table_arr, index_arr);
         jlong column_idx = index_arr[arr_len - 1];
-        TableRef table_ref = getTableByArray(nativeQueryPtr, index_arr);
+        //TableRef table_ref = getTableByArray(nativeQueryPtr, index_arr);
```
Comment (bmeike) [inline realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp]: Delete, not comment, ya?

File: realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp [hunk 13]
```diff
@@ -1704,18 +1730,18 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeIsEmpty(JNIEnv* e
...
 
         // FIXME!!!  Support a backlink as the last column in a field descriptor
```
Comment (bmeike) [inline realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp]: If it is fixed, let's remove the FIXME.  I think it is good to go, right?

File: realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp [hunk 14]
```diff
@@ -1760,6 +1786,71 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeIsEmpty(JNIEnv* e
...
+                                                   jlongArray columnIndexes, jlongArray tablePointers) {
+    JniLongArray table_arr(env, tablePointers);
+    JniLongArray index_arr(env, columnIndexes);
+    jsize arr_len = index_arr.len();
+    Query* pQuery = Q(nativeQueryPtr);
+    try {
+        TableRef src_table_ref = getTableForLinkQuery(nativeQueryPtr, table_arr, index_arr);
+        jlong column_idx = index_arr[arr_len - 1];
+
+        // FIXME!!!  Support a backlink as the last column in a field descriptor
```
Comment (bmeike) [inline realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp]: Is this just copypasta?

File: realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp [hunk 15]
```diff
@@ -1760,6 +1786,71 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeIsEmpty(JNIEnv* e
...
+                case type_Double:
+                case type_Timestamp:
+                default:
+                    ThrowException(env, IllegalArgument, "isNotEmpty() only works on String, byte[] and RealmList.");
+                    return;
+            }
+        }
+        else {
+            // Linked queries
+            // FIXME!!!  Support a backlink as an internal column in a field descriptor
```
Comment (bmeike) [inline realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp]: ... and this?

File: realm/realm-library/src/main/java/io/realm/internal/TableQuery.java [hunk 16]
```diff
@@ -396,7 +396,9 @@ public TableQuery isEmpty(long[] columnIndices, long[] tablePtrs) {
     }
 
     public TableQuery isNotEmpty(long[] columnIndices, long[] tablePtrs) {
-        return not().isEmpty(columnIndices, tablePtrs);
+        nativeIsNotEmpty(nativePtr, columnIndices, tablePtrs);
```
Comment (bmeike) [inline realm/realm-library/src/main/java/io/realm/internal/TableQuery.java]: I see what you were saying!  Got it!

File: realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsQueryTests.java [hunk 17]
```diff
@@ -30,7 +29,7 @@
 import static org.junit.Assert.assertTrue;
 import static org.junit.Assert.fail;
 
-@Ignore
+//@Ignore
 @RunWith(AndroidJUnit4.class)
 public class LinkingObjectsQueryTests extends QueryTests {
```
Comment (bmeike) [inline realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsQueryTests.java]: That might me by job.  Assigning it to Kenneth would be demonstrating that no good deed goes unpunished.

File: realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp [hunk 18]
```diff
@@ -84,12 +84,17 @@ static TableRef getTableForLinkQuery(jlong nativeQueryPtr, JniLongArray& tablesA
...
 
 // Return TableRef point to original table or the link table
-static TableRef getTableByArray(jlong nativeQueryPtr, JniLongArray& indicesArray)
+static TableRef getTableByArray(jlong nativeQueryPtr, JniLongArray& tablesArray, JniLongArray& indicesArray)
 {
     TableRef table_ref = Q(nativeQueryPtr)->get_table();
     jsize link_element_count = indicesArray.len() - 1;
     for (int i = 0; i < link_element_count; i++) {
-        table_ref = table_ref->get_link_target(size_t(indicesArray[i]));
+        if (TBL(tablesArray[i]) == nullptr) {
```
Comment (bmeike) [inline realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp]: So, I can say with absolute certainty, that there is still a boatload of use of those macros.  It is doubly confusing to have both techniques just a couple lines apart (... and I don't recall if that is the case, here).  If we are going to "move", this is one of the few cases that I would suggest we not do it piecemeal.  We should change the style, at the very least, for an entire file.

File: realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsQueryTests.java (hunk 17 above)
Comment (kneth) [inline realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsQueryTests.java]: Just think link query 😈

Comment (kneth) [inline realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsQueryTests.java]: But let me add a short paragraph to the annotation about queries.

File: realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp (hunk 18 above)
Comment (kneth) [inline realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp]: @cmelchior good point.

File: realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp (hunk 1 above)
Comment (kneth) [inline realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp]: We do throw - it is done in `isNullable()`.

File: realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp (hunk 3 above)
Comment (kneth) [inline realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp]: Forgot to remove the comment 😢

File: realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp (hunk 1 above)
Comment (kneth) [inline realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp]: We do throw - it is done in `isNullable()`.

File: realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp (hunk 15 above)
Comment (kneth) [inline realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp]: Forgot to remove :cry:

File: realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsQueryTests.java (hunk 6 above)
Comment (kneth) [inline realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsQueryTests.java]: I certainly agree, and adding an object or two with links can help a bit.

File: realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsQueryTests.java (hunk 17 above)
Comment (Zhuinden) [inline realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsQueryTests.java]: @cmelchior I would assume it works exactly like a link query, but on the objects that link to the object?

File: realm-annotations/src/main/java/io/realm/annotations/LinkingObjects.java [hunk 19]
```diff
@@ -91,6 +91,11 @@
  * assert john.dogs.size() == 2;
  * assert fido.owners.size() == 2;
  * }
+ * <p>
+ * Querying inverse relationship is like querying any {@code RealmResults}. This means that ain inverse relationship
```
Comment (Zhuinden) [inline realm-annotations/src/main/java/io/realm/annotations/LinkingObjects.java]: ain?

Comment (kneth) [inline realm-annotations/src/main/java/io/realm/annotations/LinkingObjects.java]: Damn `vi`; insert (i) n.

Comment (cmelchior) [inline realm-annotations/src/main/java/io/realm/annotations/LinkingObjects.java]: Use emacs ... oh wait, nothing is ever solved by using emacs 🙈

File: realm-annotations/src/main/java/io/realm/annotations/LinkingObjects.java [hunk 20]
```diff
@@ -91,6 +91,11 @@
...
+ * cannot be {@code null} but it can be empty (length is 0). It is possible to query field in the source class. This is
```
Comment (cmelchior) [inline realm-annotations/src/main/java/io/realm/annotations/LinkingObjects.java]: field -> fields

File: CHANGELOG.md [hunk 21]
```diff
@@ -4,6 +4,8 @@
 
 ### Enhancements
 
+* Added support for inverse relationships which can be declared by the `@LinkingObjects` annotation (#607).
```
Comment (cmelchior) [inline CHANGELOG.md]: We already had that. The queries are new 😉

Comment (kneth) [inline CHANGELOG.md]: It is new that it's out of beta.

Comment (cmelchior) [inline CHANGELOG.md]: Perhaps rephrase it to mention that then? At least I was confused about this wording, and I would still add a mention about the queries. Something like:

```
* Support for inverse relationships declared by the @LinkingObjects annotation is now considered stable.
* Added support for queries across reverse relationships (#607).
```
